File: python3/divide_two_integers.py
"""
Space   : O(1)
Time    : O(log n)
"""


class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        if dividend == 0:
            return 0

        is_neg = True if dividend < 0 else False
        is_neg ^= True if divisor < 0 else False

        dividend = abs(dividend)
        divisor = abs(divisor)

        ans, count = 0, 1
        temp = divisor
        while (divisor << 1) <= dividend:
            divisor <<= 1
            count *= 2

        while dividend >= divisor and dividend > 0:
            dividend -= divisor
            ans += count
            while dividend < divisor:
                divisor >>= 1
                count >>= 1
            if divisor < temp:
                break

        if is_neg:
            return -1 * min(ans, (2 ** 31))

        return min(ans, (2**31) - 1)

File: golang/divide_two_integers.go
package main

// Space	: O(1)
// Time		: O(log n)

func divide(dividend int, divisor int) int {
	if dividend == 0 {
		return 0
	}

	isNeg1 := dividend < 0
	isNeg2 := divisor < 0
	isNeg := isNeg1 != isNeg2

	if dividend < 0 {
		dividend *= -1
	}
	if divisor < 0 {
		divisor *= -1
	}

	ans, count, temp := 0, 1, divisor
	for (divisor << 1) <= dividend {
		divisor <<= 1
		count <<= 1
	}

	for (dividend >= divisor) && (dividend > 0) {
		dividend -= divisor
		ans += count
		for dividend < divisor {
			divisor >>= 1
			count >>= 1
		}
		if divisor < temp {
			break
		}
	}

	if isNeg {
		return -1 * Min(ans, 2147483648)
	}

	return Min(ans, 2147483647)
}
